src/ast/passes/semantic_analyser.h

```diff
@@ -156,7 +139,7 @@ class SemanticAnalyser : public Visitor {
   void binop_ptr(Binop &op);
   void binop_int(Binop &op);
   void binop_array(Binop &op);
-  void dereference_if_needed(Expression *&expr);
+  Expression *dereference_if_needed(Expression *expr);
```

Comment:
Yeah this is much better

Reply:
Per @danobi's comment, there may be some opportunities to simplify this further, but I don't want to understand much in the way of semantic changes.